src/operator/nn/concat.cc

```diff
@@ -154,18 +155,29 @@ bool ConcatType(const nnvm::NodeAttrs& attrs,
     }
   }
 
-  if (dtype == -1) {
-    LOG(FATAL) << "Not enough information to infer type in Concat.";
-    return false;
-  }
-
   size_t nin = param_.num_args;
-  in_type->clear();
-  for (size_t i = 0; i < nin; ++i) in_type->push_back(dtype);
-
-  out_type->clear();
-  out_type->push_back(dtype);
 
+  // if in types are known out types are unknown
+  if (dtype != -1 && (*out_type)[0] == -1) {
+    (*out_type)[0] = dtype;
+    in_type->clear();
+    for (size_t i = 0; i < nin; ++i) {
+      in_type->push_back(dtype);
+    }
+  // if out types are known in types are unknown
+  } else if ((*out_type)[0] != -1 && dtype == -1) {
+    in_type->clear();
+    for (size_t i = 0; i < nin; ++i) {
+      in_type->push_back((*out_type)[0]);
+    }
+  // if both out_types and in_types are known, and different
+  } else if ((*out_type)[0] != -1 && dtype != -1 && ((*out_type)[0] != dtype)) {
+    std::ostringstream os;
+    os << "Type inconsistent, Provided = "
```

Comment:
"Provided output type" maybe?
Also, typically `LOG(FATAL)` was used for such errors - is this a new preferred way?

Reply:
within InferType logic InferTypeError is thrown for example: https://github.com/apache/incubator-mxnet/blob/master/src/operator/operator_common.h#L229

I was trying to keep it consistent with this. I can change the message to Provided output type

Reply:
I have changed the error message.

Reply:
Makes sense, thanks for explanation :-)